src/filter.rs

```diff
@@ -0,0 +1,513 @@
+
+                    out.section(&out_elements);
+                    false
+                }
+
+                Payload::CodeSectionStart { count, .. } => {
+                    num_funcs = count;
+                    false
+                }
+
+                Payload::CodeSectionEntry(code) => {
+                    // Rewrite calls, ref.funcs, and return_calls
+                    // according to `func_remap`. (The latter two
+                    // become errors; intrinsics can only be used for
+                    // ordinary calls.)
+
+                    let mut locals = vec![];
+                    for local in code.get_locals_reader()? {
+                        let (count, ty) = local?;
+                        let ty = parser_to_encoder_ty(ty);
+                        locals.push((count, ty));
+                    }
+
+                    let mut func = wasm_encoder::Function::new(locals);
+                    let mut last_offset = code.range().start;
+                    let mut skip = true;
+                    for entry in code.get_operators_reader()?.into_iter_with_offsets() {
+                        let (op, offset) = entry?;
+                        if !skip {
+                            func.raw(module[last_offset..offset].iter().cloned());
+                        }
+                        last_offset = offset;
+
+                        skip = match op {
+                            wasmparser::Operator::Call { function_index } => {
+                                match self.func_remap.get(&function_index).unwrap() {
+                                    FuncRemap::Index(i) => {
+                                        func.instruction(&wasm_encoder::Instruction::Call(*i));
+                                    }
+                                    FuncRemap::InlinedBytecode(ops) => {
+                                        for op in ops {
+                                            func.instruction(op);
+                                        }
+                                    }
+                                }
+                                true
+                            }
+                            wasmparser::Operator::ReturnCall { function_index } => {
+                                match self.func_remap.get(&function_index).unwrap() {
+                                    FuncRemap::Index(i) => {
+                                        func.instruction(&wasm_encoder::Instruction::ReturnCall(
+                                            *i,
+                                        ));
+                                    }
+                                    FuncRemap::InlinedBytecode(ops) => {
+                                        for op in ops {
+                                            func.instruction(op);
+                                        }
+                                        func.instruction(&wasm_encoder::Instruction::Return);
+                                    }
```

Comment:
Ah, good catch; here I updated the comment, as (IIRC) I think I ran into an actually legitimate return call of an intrinsic when I was experimenting with `-mtail-call` and the rewrite here is valid. `ref.func` is the only remaining disallowed case (because the intrinsics are erased so one can't take a ref to them).